### PR TITLE
Fmay/support dark mode

### PR DESCRIPTION
**Pull request by travislondon**
Support Dark Mode

Current implementation looks at the current theme and
applies default colors accordingly.

NOTE: These changes do not listen for theme changes, but
we already get a notification from eclipse to restart before
changes take full effect.  Later support could be extended
to adjust when the theme changes.

Old workspaces will consider the previous defaults as user
configured due to the ui.text design.  New workspaces can be
created or in the OAL editor preferences "Restore Defaults"
can be used.

Graphics also uses a new default color scheme when dark mode
is used.

Support projects named differently than actual folder structure.

Eclipse allows you to name a project independently of the folder
on disk.  When you have this situattion Import Model Component
will create unresolvable proxies.  This can create an issue with
many areas of the tool, but the one being fixed here is that global
types are not available as the results are added to the proxy and
not the real element.  There is only one line that needs changing
to address this issue and in that case we were taking an eclipse
resource an converting it to a filesystem object.  This was unneccesary
as the full path of the resource object is the exact path we need.

### CODE REVIEW

File: src/org.xtuml.bp.core/src/org/xtuml/bp/core/util/UIUtil.java [hunk 1]
```diff
@@ -806,4 +806,29 @@ public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
 		}
 
 	}
+
+	public static boolean isDarkTheme() {
+		// get the current workbench theme, set defaults accordingly
+		// NOTE, we could import the swt css libraries but they are not
+		// considered API and access is discouraged. Instead of adding
+		// the dependencies we just use reflection, worse case is we
+		// revert to light mode
```
Comment (leviathan747) [inline src/org.xtuml.bp.core/src/org/xtuml/bp/core/util/UIUtil.java:815]: How are other plugins accessing the theme? Surely we are not the only one with custom elements that we want to follow the current theme.

File: src/org.xtuml.bp.io.core/arc/gen_import_java.inc [hunk 2]
```diff
@@ -505,7 +505,7 @@ public class ${class_name} extends CoreImport
             IPath projLoc = null;
                 projLoc = file.getProject().getLocation();
                 if (projLoc.isPrefixOf(curPath)) {
-                    projRelPath = curPath.removeFirstSegments(projLoc.segmentCount() - 1);
+                    projRelPath = file.getFullPath();
```
Comment (leviathan747) [inline src/org.xtuml.bp.io.core/arc/gen_import_java.inc:508]: This does not look related to the theme. Is this fixing a bug?

File: src/org.xtuml.bp.ui.graphics/src/org/xtuml/bp/ui/graphics/figures/ShapeImageFigure.java [hunk 3]
```diff
@@ -76,6 +76,9 @@ protected void fillShape(Graphics graphics) {
 			return;
 		}
 		Color defaultColor = specification.getInternal();
+		if(Activator.getDefault().isDarkTheme()) {
+			defaultColor = Activator.getDefault().getColor(81, 86, 88);
```
Comment (leviathan747) [inline src/org.xtuml.bp.ui.graphics/src/org/xtuml/bp/ui/graphics/figures/ShapeImageFigure.java:80]: The color is modeled and loaded into the GD_ES instances from PEI data (defined in plugin.xml). If we are going to do this, let's properly model it. Perhaps we can add a new attribute for the dark mode color and then access the correct attribute based on the current mode.

File: src/org.xtuml.bp.ui.graphics/src/org/xtuml/bp/ui/graphics/parts/DiagramEditPart.java [hunk 4]
```diff
@@ -185,7 +185,7 @@ private Color getDefaultColorFromModel() {
 		if(mdlSpec == null) {
 			return ColorConstants.white;
 		}
-		return mdlSpec.getBackground();
+		return Activator.getDefault().isDarkTheme() ? Activator.getDefault().getColor(47,47,47) : mdlSpec.getBackground();
```
Comment (leviathan747) [inline src/org.xtuml.bp.ui.graphics/src/org/xtuml/bp/ui/graphics/parts/DiagramEditPart.java:188]: Same as above comment.

**Review by leviathan747 — CHANGES_REQUESTED** [review]

**Comment by leviathan747** [comment]
Created issue: https://support.onefact.net/issues/12480

File: src/org.xtuml.bp.io.core/arc/gen_import_java.inc (hunk 2 above)
Comment (leviathan747) [reply src/org.xtuml.bp.io.core/arc/gen_import_java.inc:508]: Ok, I just saw the comments in the PR on this. Please separate into a different branch/PR and associate with the appropriate issue.

**Comment by travislondon** [comment]
I agree with the suggestions provided here, though I do not have much time to look into it at this point.  I suggest you dismiss this PR for now.  When I get a solution in place with the suggested changes I will offer a new PR.

**Comment by leviathan747** [comment]
Will do, however the change fixing the proxy issue sounds like it is worth merging now. If you peel that out into its own commit/branch, I will merge it into master.

**Pull request closed without merging**